### PR TITLE
Hardening/1384 mhd tuning

### PR DESCRIPTION
**Pull request by kzangeli**
Limit the maximum number of simultaneous connections, using the CLI option -maxConnections.
Use MHD thread pool for incoming connections, using CLI option -threadPoolSize.
Removed "main-dependency" in rest-lib.
Beautified a functest (accept_fail.test)

### CODE REVIEW

File: src/lib/rest/rest.cpp [hunk 1]
```diff
@@ -1029,8 +1035,8 @@ static int connectionTreat
     // Add to the size of the accumulated read buffer
     ciP->payloadSize += *upload_data_size;
 
-    // Zero the payload (not really necessary)
-    ciP->payload[ciP->payloadSize + 1] = 0;
+    // Zero-terminate the payload
```
Comment (fgalan) [inline src/lib/rest/rest.cpp]: Is this bugfix related with the topic of the PR? Or something that has been discovered during the process of implementing the MHD thread pool?

File: src/lib/rest/rest.cpp [hunk 2]
```diff
@@ -1138,14 +1144,22 @@ static int connectionTreat
 */
 static int restStart(IpVersion ipVersion, const char* httpsKey = NULL, const char* httpsCertificate = NULL)
 {
-  bool   mhdStartError = true;
-  size_t memoryLimit   = connectionMemory * 1024; // connectionMemory are kilobytes
+  bool      mhdStartError  = true;
+  size_t    memoryLimit    = connMemory * 1024; // connMemory is expressed in kilobytes
+  MHD_FLAG  flag           = MHD_USE_THREAD_PER_CONNECTION;
```
Comment (fgalan) [inline src/lib/rest/rest.cpp]: `flag` is rather generic as variable name... what about `connectionType` or `serverMode` or similar?

File: src/lib/rest/rest.cpp [hunk 3]
```diff
@@ -1278,14 +1303,17 @@ void restInit
...
 
-  port          = _port;
-  restServiceV  = _restServiceV;
-  ipVersionUsed = _ipVersion;
-  serveFunction = (_serveFunction != NULL)? _serveFunction : serve;
-  acceptTextXml = _acceptTextXml;
-  multitenant   = _multitenant;
-  rushHost      = _rushHost;
-  rushPort      = _rushPort;
+  port             = _port;
```
Comment (fgalan) [inline src/lib/rest/rest.cpp]: Only indent changes in this part?

File: src/lib/rest/rest.h [hunk 4]
```diff
@@ -75,14 +75,17 @@ extern void restInit
    IpVersion           _ipVersion,
    const char*         _bindAddress,
    unsigned short      _port,
-   bool                _multitenant   = false,
```
Comment (fgalan) [inline src/lib/rest/rest.h]: Only indent changes in this part?

File: test/functionalTest/cases/0000_cli/command_line_options.test [hunk 5]
```diff
@@ -62,6 +62,8 @@ Usage: contextBroker  [option '-U' (extended usage)]
                       [option '-subCacheIval' <interval in seconds between calls to Subscription Cache refresh (0: no refresh)>]
                       [option '-notificationMode' <notification mode (persistent|transient|none)>]
                       [option '-noCache' (disable subscription cache for lookups)]
-                      [option '-connectionMemory' <maximum memory size per connection in kilobytes>]
+                      [option '-connectionMemory' <maximum memory size per connection (in kilobytes)>]
+                      [option '-maxConnections' <maximum number of simultaneous connections>]
+                      [option '-threadPoolSize' <size of thread pool for incoming connections>]
```
Comment (fgalan) [inline test/functionalTest/cases/0000_cli/command_line_options.test]: cli.md should include entries describing the new flags.

File: test/functionalTest/cases/0000_content_related_headers/accept_fail.test [hunk 6]
```diff
@@ -25,12 +25,32 @@ Accept header fail
 
 --SHELL-INIT--
 dbInit CB
-brokerStart CB
+brokerStart CB 38,39
```
Comment (fgalan) [inline test/functionalTest/cases/0000_content_related_headers/accept_fail.test]: Only style/cosmetica changes in this .test?

File: src/lib/rest/rest.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: I found this bug the first time I turned the MHD thread pool on

File: src/lib/rest/rest.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: Sure, "flag" is the MHD word for this, but yes, it **is** extremely generic

File: src/lib/rest/rest.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: Almost never *only* indents.
The indents nirmally come as a new var has been introduced.

In this case, connMemory, maxConns and threadPoolSize

File: src/lib/rest/rest.h (hunk 4 above)
Comment (kzangeli) [inline src/lib/rest/rest.h]: same same

File: test/functionalTest/cases/0000_cli/command_line_options.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/0000_cli/command_line_options.test]: Ah, that I have totally forgotten about .. :-(
Coming ...

File: test/functionalTest/cases/0000_content_related_headers/accept_fail.test (hunk 6 above)
Comment (kzangeli) [inline test/functionalTest/cases/0000_content_related_headers/accept_fail.test]: Yes.
This test failed (that's how I discovered the bug in connectionTreat.
In order toi understandthe test I had to 'transform' it to something readable

File: src/lib/rest/rest.cpp (hunk 1 above)
Comment (fgalan) [inline src/lib/rest/rest.cpp]: Bugfix worth to be mentioned in CNR.

File: src/lib/rest/rest.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: Fixed in 78ee8ac

File: test/functionalTest/cases/0000_cli/command_line_options.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/0000_cli/command_line_options.test]: Fixed in 78ee8ac

File: src/lib/rest/rest.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: Fixed in 78ee8ac

File: src/lib/rest/rest.cpp (hunk 3 above)
Comment (fgalan) [inline src/lib/rest/rest.cpp]: NTC

File: src/lib/rest/rest.h (hunk 4 above)
Comment (fgalan) [inline src/lib/rest/rest.h]: NTC

File: test/functionalTest/cases/0000_content_related_headers/accept_fail.test (hunk 6 above)
Comment (fgalan) [inline test/functionalTest/cases/0000_content_related_headers/accept_fail.test]: NTC

File: src/lib/rest/rest.cpp [hunk 7]
```diff
@@ -1168,6 +1182,8 @@ static int restStart(IpVersion ipVersion, const char* httpsKey = NULL, const cha
                                    MHD_OPTION_HTTPS_MEM_KEY,            httpsKey,
                                    MHD_OPTION_HTTPS_MEM_CERT,           httpsCertificate,
                                    MHD_OPTION_CONNECTION_MEMORY_LIMIT,  memoryLimit,
+                                   MHD_OPTION_CONNECTION_LIMIT,         maxConns,
+                                   MHD_OPTION_THREAD_POOL_SIZE,         threadPoolSize,
```
Comment (fgalan) [inline src/lib/rest/rest.cpp]: What is the effect of the MHD_OPTION_THREAD_POOL_SIZE flag when MHD_USE_THREAD_PER_CONNECTION is used? Is it ignored?

File: test/unittests/main_UnitTest.cpp [hunk 8]
```diff
@@ -67,7 +67,6 @@ unsigned int  cprForwardLimit       = 1000;
 bool          noCache               = false;
 char          fwdHost[64];
 char          notificationMode[64];
-unsigned int  connectionMemory      = 64;
```
Comment (fgalan) [inline test/unittests/main_UnitTest.cpp]: Why this is no loner used in unit test? Maybe it has been moved to a library?

File: src/lib/rest/rest.cpp (hunk 7 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: Nope, it would give an error (if pool size > 1).
However, this never happens as:

    if (threadPoolSize != 0)
    {
      serverMode = MHD_USE_SELECT_INTERNALLY;
      ...

Tested empirically. It would be nice to have a make step to run all functests with/without MHD thread pool.
I have done this, once, manually but it wouldn't hurt to have it automatized ...

File: test/unittests/main_UnitTest.cpp (hunk 8 above)
Comment (kzangeli) [inline test/unittests/main_UnitTest.cpp]: This here was a big mistake. The library depended on a global variable of the executable.
Now, this value is passed to restInit, the way it *must* be. The new variable holding this value is inside the rest library.

File: src/lib/rest/rest.cpp (hunk 7 above)
Comment (fgalan) [inline src/lib/rest/rest.cpp]: What it should be great is to have a set of env vars to model different "working models", e.g. w/ or w/o cache, w/ or w/o threadd pool, with the different notification modes, etc. and configure jenkins to do a pass in all them.

Issue about it is on the way...

File: test/unittests/main_UnitTest.cpp (hunk 8 above)
Comment (fgalan) [inline test/unittests/main_UnitTest.cpp]: NTC

File: src/lib/rest/rest.cpp (hunk 7 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: We could add an option about 'working model' (which gives a env var at the same time).
We would need to compare it to the other options though

Comment (kzangeli) [inline src/lib/rest/rest.cpp]: Explaining comment added in 41676e4

**Comment by fgalan** [comment]
LGTM
